Mage/src/main/java/mage/abilities/effects/common/GainsChoiceOfAbilitiesEffect.java

```diff
@@ -0,0 +1,120 @@
+
+package mage.abilities.effects.common;
+
+import mage.abilities.Ability;
+import mage.abilities.Mode;
+import mage.abilities.effects.OneShotEffect;
+import mage.abilities.effects.common.continuous.GainAbilityTargetEffect;
+import mage.choices.Choice;
+import mage.choices.ChoiceImpl;
+import mage.constants.Duration;
+import mage.constants.Outcome;
+import mage.game.Game;
+import mage.game.permanent.Permanent;
+import mage.players.Player;
+import mage.target.targetpointer.FixedTargets;
+
+import java.util.*;
+
+/**
+ * Something gains one ability of the player's choice from a list of abilities until end of turn.
+ * Affects any targets set and optionally the source as well. Have no targets to affect only the source.
+ *
+ * @author notgreat, TheElk801
+ */
+public class GainsChoiceOfAbilitiesEffect extends OneShotEffect {
+
+    private final Map<String, Ability> abilityMap = new LinkedHashMap<>();
+    private final boolean affectSource, includeEnd;
+    private final String targetDescription;
+
+    public GainsChoiceOfAbilitiesEffect(Ability... abilities) {
+        this(false, null, true, abilities);
+    }
+
+    public GainsChoiceOfAbilitiesEffect(boolean affectSource, Ability... abilities) {
+        this(affectSource, null, true, abilities);
+    }
+
+    public GainsChoiceOfAbilitiesEffect(boolean affectSource, String targetDescription, boolean includeEnd, Ability... abilities) {
+        super(Outcome.AddAbility);
+        this.affectSource = affectSource;
+        this.targetDescription = targetDescription;
+        this.includeEnd = includeEnd;
+        for (Ability ability : abilities) {
+            this.abilityMap.put(ability.getRule(), ability);
+        }
+    }
+
+    protected GainsChoiceOfAbilitiesEffect(final GainsChoiceOfAbilitiesEffect effect) {
+        super(effect);
+        this.affectSource = effect.affectSource;
+        this.abilityMap.putAll(effect.abilityMap);
```

Comment:
Not critical here, but it better to use CardUtil.deepCopy to copy field collections. Don’t use shared refs.